src/token/colmetadata-token-parser.ts

```diff
@@ -60,57 +69,256 @@ function readColumnName(parser: Parser, options: InternalConnectionOptions, inde
+                  // console.log('>> read CEK value done')
+
+                  callback();
+                });
+              });
+            });
+          });
+        });
+      });
+    });
+  });
+
+  // console.log('>> ??? ', i, parser.position)
+}
 
+function readCEKTableCb(columnCount: number, parser: Parser, resolved: { val: boolean; }, resolve: (arg0: ColMetadataToken) => void, cekList?: CEKEntry[]) {
+  // console.log('>> entering readCEKTableCb, parser.position = ', parser.position, ', parser.buffer.length = ', parser.buffer.length)
   const columns: ColumnMetadata[] = [];
   for (let i = 0; i < columnCount; i++) {
+    // console.log(' i = ', i, ' (', columnCount, ')', ', <', parser.position, '>', ', [', parser.buffer.length, ']')
     let column: ColumnMetadata;
 
-    readColumn(parser, parser.options, i, (c) => {
+    readColumn(parser, parser.options, i, cekList, (c) => {
       column = c;
     });
 
-    while (parser.suspended) {
-      await parser.streamBuffer.waitForChunk();
-
-      parser.suspended = false;
-      const next = parser.next!;
-
-      next();
+    if (!column!) {
+      // console.log('>> Column is ', column!, ', breaking!');
+      break;
     }
+    else {
+      columns.push(column!);
+      if (i === columnCount - 1) {
+        // console.log('>> RESOLVING colmetdata token ')
+        resolved.val = true;
+        // console.log('>>>>> resolved? ', resolved)
+        return resolve(new ColMetadataToken(columns));
+      }
+    }
+  }
+}
 
-    columns.push(column!);
+async function colMetadataParser(parser: Parser): Promise<ColMetadataToken> {
+  while (parser.buffer.length - parser.position < 2) {
+    await parser.streamBuffer.waitForChunk();
   }
 
-  return new ColMetadataToken(columns);
+  const columnCount = parser.buffer.readUInt16LE(parser.position);
+  parser.position += 2;
+
+  return new Promise((resolve) => {
+    const parser_starting_position = parser.position;
+    let resolved = { val: false };
+    let whileLoopPaused = false;
+    readCEKTable(parser, parser.options, (cekList?: CEKEntry[]) => {
+      readCEKTableCb(columnCount, parser, resolved, resolve, cekList);
+    });
+    // console.log('>>>> resolved val = ', resolved);
+    if (!resolved.val) {
+      // console.log('>> NOT Resolved')
+      while (parser.suspended && !whileLoopPaused) {
+        // console.log('parser suspended!')
+        parser.position = parser_starting_position;
+        whileLoopPaused = true;
+        parser.streamBuffer.waitForChunk().then(() => {
+          // console.log('>> streambuffer data loadedd')
+          parser.suspended = false;
+          whileLoopPaused = false;
+          readCEKTable(parser, parser.options, (cekList?: CEKEntry[]) => {
+            readCEKTableCb(columnCount, parser, resolved, resolve, cekList);
+          });
+        })
+      }
+    }
+  })
```

Comment:
@arthurschreiber I added these changes while merging this branch with the latest master. Let me know what you think. These changes are to accommodate when there is not enough column metadata buffer and needs to wait for more ( Our original implementation didn't' take this scenario into account). The tests `always-encrypted/collation-support-test` and `always-encrypted/cek-cache-test` trigger this scenario).